logtail/logtail.go

```diff
@@ -573,169 +596,211 @@ func (l *Logger) sendLocked(jsonBlob []byte) (int, error) {
-		case '\n':
-			b = append(b, '\\', 'n')
-		case '\r':
-			b = append(b, '\\', 'r')
-		case '\t':
-			b = append(b, '\\', 't')
-		case '"':
-			b = append(b, '\\', '"')
-		case '\\':
-			b = append(b, '\\', '\\')
-		default:
-			// TODO: what about binary gibberish or non UTF-8?
-			b = append(b, c)
-		}
+
+	return dst
+}
+
+// appendText appends a raw text message in the Tailscale JSON log entry format.
+func (l *Logger) appendText(dst, src []byte, skipClientTime bool, procID uint32, procSequence uint64, level int) []byte {
+	dst = slices.Grow(dst, len(src))
+	dst = append(dst, '{')
+	dst = l.appendMetadata(dst, skipClientTime, false, procID, procSequence, "", nil, level)
+	if len(src) == 0 {
+		dst = bytes.TrimRight(dst, ",")
+		return append(dst, "}\n"...)
 	}
-	if nTruncated > 0 {
-		b = append(b, "…+"...)
-		b = strconv.AppendInt(b, int64(nTruncated), 10)
+
+	// Append the text string, which may be truncated.
+	// Invalid UTF-8 will be mangled with the Unicode replacement character.
+	max := 16 << 10
+	if l.lowMem {
+		max = 4 << 10
 	}
-	b = append(b, "\"}\n"...)
-	return b
+	dst = append(dst, `"text":`...)
+	dst = appendTruncatedString(dst, src, max)
+	return append(dst, "}\n"...)
+}
+
+// appendTruncatedString appends a JSON string for src,
+// truncating the src to be no larger than n.
+func appendTruncatedString(dst, src []byte, n int) []byte {
+	srcLen := len(src)
+	src = truncate.String(src, n)
+	dst, _ = jsontext.AppendQuote(dst, src) // ignore error; only occurs for invalid UTF-8
+	if srcLen > len(src) {
+		dst = dst[:len(dst)-len(`"`)] // trim off preceding double-quote
+		dst = append(dst, "…+"...)
+		dst = strconv.AppendInt(dst, int64(srcLen-len(src)), 10)
+		dst = append(dst, '"') // re-append succeeding double-quote
+	}
+	return dst
+}
+
+func (l *Logger) AppendTextOrJSONLocked(dst, src []byte) []byte {
+	l.clock = tstime.StdClock{}
+	return l.appendTextOrJSONLocked(dst, src, 0)
 }
 
-func (l *Logger) encodeLocked(buf []byte, level int) []byte {
+// appendTextOrJSONLocked appends a raw text message or a raw JSON object
+// in the Tailscale JSON log format.
+func (l *Logger) appendTextOrJSONLocked(dst, src []byte, level int) []byte {
 	if l.includeProcSequence {
 		l.procSequence++
 	}
-	if buf[0] != '{' {
-		return l.encodeText(buf, l.skipClientTime, l.procID, l.procSequence, level) // text fast-path
+	if len(src) == 0 || src[0] != '{' {
+		return l.appendText(dst, src, l.skipClientTime, l.procID, l.procSequence, level)
 	}
 
-	now := l.clock.Now()
-
-	obj := make(map[string]any)
-	if err := json.Unmarshal(buf, &obj); err != nil {
-		for k := range obj {
-			delete(obj, k)
+	// Check whether the input is a valid JSON object and
+	// whether it contains the reserved "logtail" name at the top-level.
+	var logtailKeyOffset, logtailValOffset, logtailValLength int
+	validJSON := func() bool {
+		// TODO(dsnet): Avoid allocation of bytes.Buffer struct.
+		dec := &l.jsonDec
+		dec.Reset(bytes.NewBuffer(src))
```

Comment:
nit: you don't actually need a Buffer here, you could just use `bytes.NewReader`

Reply:
The v2 json.Decoder has an internal optimization where it looks directly at the underlying `[]byte` for `bytes.Buffer`. It can't do this same optimization for `bytes.Reader` since the underlying `[]byte` is somewhat hidden from the API.

Reply:
(This is only allocating the `bytes.Buffer` struct itself, but reusing the provided `[]byte`). There's an allocation going on, but it's an order-of-magnitude less than before.

Reply:
> It can't do this same optimization for bytes.Reader since the underlying []byte is somewhat hidden from the API.

Kudos to you for seeing that, and also yuk.

> This is only allocating the bytes.Buffer struct itself

Yeah, I wasn't concerned about allocations, more just about using a seemingly more expansive implementation than necessary. Might be worth a comment.